### DISCUSSION TITLE
Identifying Custom vs Standard PMD Rule

### DISCUSSION THREAD

**Discussion by BrundaSreedhar** [pmd/pmd#4662] [Q&A]
How do you identify if a PMD rule is a custom rule or a standard rule and get the category for the same?

**Comment by adangel** [comment]
Given a ruleset, you can assume, that any rule-ref that starts with `category/<language>/<category.xml>` is a standard rule. See why: https://docs.pmd-code.org/latest/pmd_userdocs_making_rulesets.html#referencing-a-single-rule

If you have only the rule name, you can search for it in the documentation: https://docs.pmd-code.org/latest/ (searchbox in the top bar, just type)
When you find something, then it is a standard rule. You'll see in the menu, in which category you are.

I don't think, that there is currently a easy way to automate this (refs #4295 and #1335).

**Reply by anmol-copado** [reply]
I struggle to detect the rule type for cases where rule-ref is pointing to a custom rule file.

Example: https://github.com/rsoesemann/unhappy-soup/blob/master/pmd/ruleset.xml#L45

In my current implementation, I add an extra check to see if the rule ref is a file in the current folder. But this makes my application a bit slow and I was wondering if there is any easy way

**Reply by rsoesemann** [reply]
@anmol-copado I know this is not elegant but faster. Why don't you keep a List of Standard Rule API names in your app instead of making this decision every time. For your you need to update this when you update your version of PMD but that's easy compared to crawling the docs.

**Reply by anmol-copado** [reply]
Thanks Robert! 
I like this approach
